edb/server/service_manager.py

```diff
@@ -0,0 +1,223 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2016-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import annotations
+from typing import *
+
+import logging
+import os
+import socket
+import sys
+
+
+SD_LISTEN_FDS_START = 3
+
+
+logger = logging.getLogger('edb.server')
+
+
+def _stream_socket_from_fd(fd: int) -> Optional[socket.socket]:
+    try:
+        sock = socket.socket(fileno=fd)
+    except OSError:
+        logger.warning(
+            f"activation file descriptor {fd} is not a socket "
+            f", ignoring"
+        )
+        return None
+
+    if sock.family not in {socket.AF_INET, socket.AF_INET6}:
+        logger.warning(
+            f"activation file descriptor {fd} is not an AF_INET[6] socket "
+            f", ignoring"
+        )
+        return None
+
+    if sock.type != socket.SOCK_STREAM:
+        logger.warning(
+            f"activation file descriptor {fd} is not an SOCK_STREAM "
+            f"socket, ignoring"
+        )
+        return None
+
+    return sock
+
+
+def sd_notify(message: str) -> None:
+    notify_socket = os.environ.get('NOTIFY_SOCKET')
+    if not notify_socket:
+        return
+
+    if notify_socket[0] == '@':
+        notify_socket = '\0' + notify_socket[1:]
+
+    sd_sock = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
+
+    try:
+        sd_sock.connect(notify_socket)
+        sd_sock.sendall(message.encode())
+    except Exception as e:
+        logger.info('Could not send systemd notification: %s', e)
+    finally:
+        sd_sock.close()
+
+
+def sd_get_activation_listen_sockets(
+    names: list[str],
+) -> dict[str, socket.socket]:
+    listen_pid = os.environ.pop("LISTEN_PID", "")
```

Comment:
Why `pop` and not a simple `get`? Can make it harder to debug.

Reply:
This is intentional and recommended to avoid child processes from thinking they are being socket-activated.

Reply:
Add a comment?

Reply:
Apparently you're also supposed to set FD_CLOEXEC for any passed FDs + support SD_LISTEN_FDS_START?